File: problems/2302/solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    long long count(long long length)
    {
        return (1 + length) * length / 2;
    }
    long long countSubarrays(vector<int> &nums, long long k)
    {
        int ptr1 = 0, ptr2 = 0, last_ptr2 = -1;
        long long current_sum = nums[0], result = 0;
        while (ptr2 <= nums.size())
        {
            if (ptr2 == nums.size())
            {
                result += count(ptr2 - ptr1);
                if (last_ptr2 >= ptr1)
                {
                    result -= count(last_ptr2 - ptr1 + 1);
                }
                break;
            }
            if (current_sum * (ptr2 - ptr1 + 1) >= k)
            {
                result += count(ptr2 - ptr1);
                if (last_ptr2 >= ptr1)
                {
                    result -= count(last_ptr2 - ptr1 + 1);
                }
                last_ptr2 = ptr2 - 1;
                if (ptr1 == ptr2)
                {
                    ptr2 += 1;
                    if (ptr2 < nums.size())
                    {
                        current_sum += nums[ptr2];
                    }
                }
                current_sum -= nums[ptr1];
                ptr1 += 1;
            }
            else
            {
                ptr2 += 1;
                if (ptr2 < nums.size())
                {
                    current_sum += nums[ptr2];
                }
            }
        }
        return result;
    }
};

File: problems/2302/solution.py
from typing import List

class Solution:
    def count(self, length: int) -> int:
        return (1 + length) * length // 2
    def countSubarrays(self, nums: List[int], k: int) -> int:
        ptr1 = 0
        ptr2 = 0
        last_ptr2 = -1
        result = 0
        current_sum = nums[0]
        while ptr2 <= len(nums):
            if ptr2 == len(nums):
                result += self.count(ptr2 - ptr1) 
                if last_ptr2 >= ptr1:
                    result -= self.count(last_ptr2 - ptr1 + 1)
                break
            if current_sum * (ptr2 - ptr1 + 1) >= k:
                result += self.count(ptr2 - ptr1) 
                if last_ptr2 >= ptr1:
                    result -= self.count(last_ptr2 - ptr1 + 1)
                last_ptr2 = ptr2 - 1
                if ptr1 == ptr2:
                    ptr2 += 1
                    if ptr2 < len(nums):
                        current_sum += nums[ptr2]
                current_sum -= nums[ptr1]
                ptr1 += 1
            else:
                ptr2 += 1
                if ptr2 < len(nums):
                    current_sum += nums[ptr2]
        return result


if __name__ == "__main__":
    nums = [2,1,4,3,5]
    k = 10
    print(Solution().countSubarrays(nums=nums,k=k))